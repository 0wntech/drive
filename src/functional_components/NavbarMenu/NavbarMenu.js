import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import styles from './NavbarMenu.module.scss';
import DropdownMenu from '../DropdownMenu/DropdownMenu';
import { setCurrentPath } from '../../actions/appActions';
import { logout } from '../../actions/userActions';
import { getUsernameFromWebId } from '../../utils/url';
import { navigate, getInitialsFromUser } from '../../utils/helper';
import useWindowDimension from '../../hooks/useWindowDimension';
import styleConstants from '../../styles/constants.scss';
import SvgSettings from '../../assets/svgIcons/Settings';
import DefaultIcon from '../DefaultIcon/DefaultIcon';

export const NavbarMenu = ({
    className,
    history,
    logout,
    user,
    rootUrl,
    setCurrentPath,
    resetError,
    dispatch,
}) => {
    const [isDropdownExpanded, setDropdownExpanded] = useState(false);
    const [profilePictureError, setProfilePictureError] = useState(false);
    const profilePictureRef = useRef();

    useEffect(() => setProfilePictureError(false), [user]);

    const DROPDOWN_OPTIONS = [
        {
            onClick: () => {
                navigate('/home', history, dispatch);
                setCurrentPath(rootUrl);
            },
            label: 'Home',
        },
        { onClick: () => history.push('/settings'), label: 'Settings' },
        {
            onClick: () => navigate('/profile', history, dispatch, resetError),
            label: 'Profile',
        },
        {
            onClick: () => navigate('/apps', history, dispatch, resetError),
            label: 'Apps',
        },
        { onClick: () => logout(), label: 'Logout' },
    ];

    // eslint-disable-next-line no-unused-vars
    const { _, width } = useWindowDimension();

    const getMenuHead = () => (
        <div className={styles.profileSection}>
            {width < styleConstants.screen_l &&
            history.location.pathname === '/profile' ? (
                <div
                    className={styles.settings}
                    onClick={(e) => {
                        e.stopPropagation();
                        setDropdownExpanded(false);
                        navigate('/settings', history, dispatch, resetError);
                    }}
                >
                    <SvgSettings />
                </div>
            ) : user && user.picture && !profilePictureError ? (
                <div
                    className={styles.profileIcon}
                    ref={profilePictureRef}
                    onClick={(e) => {
                        e.stopPropagation();
                        setDropdownExpanded(false);
                        navigate('/profile', history, dispatch, resetError);
                    }}
                    data-test-id="navigation-profile-picture"
                >
                    <img
                        alt={`${getUsernameFromWebId(user.webId)}'s profile`}
                        onLoad={() => {
                            profilePictureRef.current.style.backgroundImage = `url(${user.picture})`;
                        }}
                        onError={() => setProfilePictureError(true)}
                        src={user.picture}
                    />
                </div>
            ) : (
                <DefaultIcon
                    onClick={(e) => {
                        e.stopPropagation();
                        setDropdownExpanded(false);
                        navigate('/profile', history, dispatch, resetError);
                    }}
                    className={styles.defaultProfileIcon}
                    initials={getInitialsFromUser(user)}
                />
            )}

            <div data-test-id="navbar-username" className={styles.username}>
                {user.name ? user.name : getUsernameFromWebId(user.webId)}
            </div>
        </div>
    );

    return (
        <div className={className}>
            {user && user.webId ? (
                <DropdownMenu
                    menuHead={getMenuHead()}
                    options={DROPDOWN_OPTIONS}
                    isExpanded={isDropdownExpanded}
                    setExpanded={setDropdownExpanded}
                />
            ) : null}
        </div>
    );
};

const mapStateToProps = (state) => ({
    user: state.user.user,
    rootUrl: state.app.rootUrl,
});

NavbarMenu.propTypes = {
    className: PropTypes.string,
    history: PropTypes.object,
    user: PropTypes.object,
    logout: PropTypes.func,
};

export default withRouter(
    connect(mapStateToProps, (dispatch) => ({
        ...bindActionCreators({ logout, setCurrentPath }, dispatch),
        dispatch,
    }))(NavbarMenu)
);
