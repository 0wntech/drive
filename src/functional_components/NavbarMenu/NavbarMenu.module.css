.dropdown a {
    padding: inherit 0;
    width: 100%;
}

.profileIcon {
    min-height: 56px;
    min-width: 56px;
    background-size: cover;
    background-position: center;
    border-radius: 50%;
    cursor: pointer;
    margin-left: 18px;
}

.profileSection {
    display: flex;
    align-items: center;
    height: 100%;
    width: 80%;
}

.username {
    display: table-cell;
    vertical-align: middle;
    letter-spacing: 0.2em;
    max-height: 63px;
    width: 100%;
    line-height: 21px;
    font-size: 16px;
    font-family: 'Helvetica Neue Light';
    margin: 0 10px;
    text-overflow: ellipsis;
    overflow: hidden;
}

.mask {
    border-radius: 20px;
    -webkit-z-index: -1;
    -moz-z-index: -1;
    -ms-z-index: -1;
    -o-z-index: -1;
    z-index: -1;
    position: absolute;
    width: 100%;
    background: #ffffff;
    height: 100%;
}

.mask.active {
    border-radius: 20px 20px 0px 20px;
}
