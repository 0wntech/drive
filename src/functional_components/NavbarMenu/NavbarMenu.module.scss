@import '../../styles/mixins.scss';
@import '../../styles/constants.scss';

.dropdown a {
    padding: inherit 0;
    width: 100%;
}

.profileIcon {
    min-height: 56px;
    min-width: 56px;
    background-size: cover;
    background-position: center;
    border-radius: 50%;
    cursor: pointer;
}

.profileSection {
    display: flex;
    align-items: center;
    height: 100%;
    width: 80%;
}

.username {
    letter-spacing: 0.2em;
    max-height: 63px;
    width: 100%;
    font-size: 16px;
    font-family: 'Helvetica Neue Light';
    margin: 0 10px;
    @include multiLineEllipsis(
        $lineHeight: 21px,
        $lineCount: 2,
        $bgColor: white
    );
}

.mask {
    border-radius: 20px;
    -webkit-z-index: -1;
    -moz-z-index: -1;
    -ms-z-index: -1;
    -o-z-index: -1;
    z-index: -1;
    position: absolute;
    width: 100%;
    background: #ffffff;
    height: 100%;
}

.mask.active {
    border-radius: 20px 20px 0px 20px;
}

@media (max-width: $screen_m) {
    .username {
        display: none;
    }

    .profileSection {
        height: initial;
        width: 100%;
    }

    .profileIcon {
        min-height: 50px;
        min-width: 50px;
    }
}
