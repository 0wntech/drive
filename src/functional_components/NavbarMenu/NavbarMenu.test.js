/* eslint-disable quotes */
import React from 'react';
import { NavbarMenu } from './NavbarMenu';
import TestRenderer from 'react-test-renderer';
import { user } from '../../test/fixtures/user';

describe('NavbarMenu', () => {
    test('Should render correctly', () => {
        const className = 'test';
        const wrapper = TestRenderer.create(
            <NavbarMenu className={className} user={user} />
        );
        expect(wrapper.toJSON()).toMatchSnapshot();
    });
    test('Should render login if no webId', () => {
        const className = 'test';

        const wrapper = TestRenderer.create(
            <NavbarMenu className={className} user={null} />
        );
        expect(wrapper.toJSON()).toMatchSnapshot();
    });
});
