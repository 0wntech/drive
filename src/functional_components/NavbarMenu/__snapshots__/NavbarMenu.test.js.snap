// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavbarMenu Should render correctly 1`] = `
<div
  className="test"
>
  <div
    className="dropdownWrapper"
  >
    <div
      className="container"
    >
      <div
        className="head"
      >
        <div
          className="profileSection"
        >
          <div
            className="profileIcon"
            data-test-id="navigation-profile-picture"
            onClick={[Function]}
            style={
              Object {
                "backgroundImage": "url('https://bejow.owntech.de/profile/meme_ben_gang.jpg')",
              }
            }
          />
          <div
            className="username"
            data-test-id="navbar-username"
          >
            Ben Wetzel
          </div>
        </div>
        <div
          className="iconWrapper"
        >
          <svg
            className="icon"
            fill="none"
            height={10}
            onClick={[Function]}
            width={14}
          >
            <path
              d="M0 0h14L7 10 0 0z"
              fill="#C4C4C4"
            />
          </svg>
        </div>
      </div>
      <div
        className="dropdownContent hide"
      >
        <div
          className="option"
          onClick={[Function]}
        >
          Home
        </div>
        <div
          className="option"
          onClick={[Function]}
        >
          Profile
        </div>
        <div
          className="option"
          onClick={[Function]}
        >
          Settings*
        </div>
        <div
          className="option"
          onClick={[Function]}
        >
          Notifications*
        </div>
        <div
          className="option"
          onClick={[Function]}
        >
          Contacts
        </div>
        <div
          className="option"
          onClick={[Function]}
        >
          Apps
        </div>
        <div
          className="option"
          onClick={[Function]}
        >
          Logout
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`NavbarMenu Should render login if no webId 1`] = `
<div
  className="test"
>
  <div
    className="container green sm"
    onClick={[Function]}
  >
    Register
  </div>
</div>
`;
